### DISCUSSION TITLE
How to unit test _bundle_ controllers and commands

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#57600] [Q&A]
Can someone point me to documentation or an example of how to unit test a controller and a command from within a bundle?

Most of the instructions are for how to unit test an _application_, but I have a controller and a command within a bundle and I'm not sure how to configuring twig, routing, etc.  

I realize controllers within bundles aren't that common, and they're aren't even enabled automatically unless you write a recipe that creates a routing config.  Nonetheless, I find it helpful to have controllers and commands within my bundles, and am finally learning how to write unit tests.  

Thanks.

**Comment by alexandre-le-borgne** [comment]
I think this answer might satisfy you : https://stackoverflow.com/a/10132372/4864628



> Normally, your controller plugs together different objects and connects them in the correct order. Maybe he calls a repository, reads some objects and returns them through the render method. Maybe he calls some other Handlers/Managers who do stuff.
> 
> This means that a controller is a high level component. More often than not this indicates that functional tests are in order instead of unit tests. You should not aim to get 100% code coverage with your unit tests. Maybe you can think of it that way: If you unit test everything the controller calls (model, validation, form, repository), what could go wrong? Most of the time it's something you only observe when using all the real classes involved when in production.
> 
> I want also like to point out that TDD does not mean that everything has to be unit-tested. It's ok to have some functional tests for the high-level code. As said, if you test the low-level components with unit-tests you should only test how they are working together which you cannot test with mocks because you tell the mocks what the return value is.
> 
> If your controller does more than plugging parts of the system together you should think about refactoring the stuff into more low-level classes you can test with unit-tests.
> 
> So my suggestion would be to use functional tests to test your controllers and use unit-tests to test your models and your business logic stuff.
> 
> If you struggle with functional tests, you can read the following:
> 
> http://symfony.com/doc/current/book/testing.html#functional-tests
> http://blog.sznapka.pl/fully-isolated-tests-in-symfony2/


If you still want to test your controller alone, I imagine you can forge a request and then create and invoke the controller action you want to test, then check the content of the response

**Reply by tacman** [reply]
It's quite a bit more work than testing a controller from within an application, I'm hoping that there's something equivalent.  Sometimes I start off writing tests from within the application, even if they're testing the bundle.  Now I want to move that test to the bundle, and am stuck on the routing and such.

I understand that I can treat the Controller as a service, and inject all the dependencies, but if there's an easier way I'll use that.

I've never had much luck with the phpunit-bridge, (vendor/bin/simple-phpunit), but maybe that's part of the solution.

**Comment by mbabker** [comment]
If your intent is to do unit tests, where you're testing something in isolation with mocks and whatnot, there is nothing special you need to do.  It's the same setup as unit testing any other piece of code.

You can use https://symfony.com/doc/current/console.html#testing-commands to help set up tests for your commands without a kernel (the docs example is written for applications with a kernel since most of the docs are focused on "how to do this in an application", but replace `$command = $application->find($commandName);` with `$command = new $commandClass;` and you're basically good to go without needing the `KernelTestCase`.

If your intent is to do functional tests that need to use the `KernelTestCase` or `WebTestCase`, that isn't documented well, but the absolute minimum requirement is that you have a test kernel to work with.  There's no One Right Way™ to accomplish this because everyone has their own testing requirements.  Remember, most applications are going to have a single Symfony version with a single set of bundles and a single configuration tree they're working with.  Whereas with many bundles, they need to be able to do tests across many Symfony versions (where configuration nodes may have been added or removed depending on what you're working with from the framework), potentially with other bundles enabled or disabled (for optional feature integrations) or different configuration trees covering different features (i.e. the DoctrineBundle has two test kernel classes with base configs running with and without the ORM installed/enabled).

**Reply by tacman** [reply]
Thanks, between your answer and @alexandre-le-borgne , I made some progress.  I've injected the controller as a service, but it's failing when I call render:

```
LogicException: You cannot use the "render" method if the Twig Bundle is not available. Try running "composer require symfony/twig-bundle".
/home/tac/g/survos/survos/packages/pixie-bundle/vendor/symfony/framework-bundle/Controller/AbstractController.php:418
```

```php
        return $this->render('@SurvosPixie/pixie/index.html.twig', [
            'dir' => $this->pixieService->getConfigDir(),
            'configs' => $configs,
        ]);
```

Even if I have twig bundle installed.
```php
        /** @var PixieController $controller */
        $controller = static::getContainer()->get(PixieController::class);
        $response = $controller->browsePixies();
        self::assertSame(200, $response->getStatusCode());
```

**Reply by stof** [reply]
Is TwigBundle registered in the kernel you use for your test ? Installing the bundle is not enough if you don't register it (for a custom kernel used as part of your bundle testsuite, Flex won't be managing the registration automatically)

**Reply by tacman** [reply]
Indeed, I had not, but I'm getting so much closer, thanks!  I wired a few more things, bu I'm still missing something in within twig, as I now get


 ✘ Controller
   ┐
   ├ Twig\Error\SyntaxError: Unknown "path" function. Did you mean "relative_path"?
   │
   │ /home/tac/g/survos/survos/packages/pixie-bundle/templates/pixie/index.html.twig:19

for the very vanilla

```twig
                        <a href="{{ path('pixie_homepage', {pixieCode: p}) }}">Show</a>
```

I tried to simply inject Twig\Environment in TestKernel.php like I did with the TwigBundle, but it's expecting a $loader that that didn'twork.
